rpc/client/ws_client.go

```diff
@@ -0,0 +1,109 @@
+package rpcclient
+
+import (
+	"encoding/json"
+	"net/http"
+	"strings"
+
+	"github.com/tendermint/tendermint/Godeps/_workspace/src/github.com/gorilla/websocket"
+	. "github.com/tendermint/tendermint/common"
+	_ "github.com/tendermint/tendermint/config/tendermint_test"
+	"github.com/tendermint/tendermint/rpc/types"
+)
+
+const wsEventsChannelCapacity = 10
+const wsResponsesChannelCapacity = 10
+
+type WSClient struct {
+	QuitService
+	Address string
+	*websocket.Conn
+	EventsCh    chan rpctypes.RPCEventResult
+	ResponsesCh chan rpctypes.RPCResponse
+}
+
+// create a new connection
+func NewWSClient(addr string) *WSClient {
+	wsClient := &WSClient{
+		Address:     addr,
+		Conn:        nil,
+		EventsCh:    make(chan rpctypes.RPCEventResult, wsEventsChannelCapacity),
+		ResponsesCh: make(chan rpctypes.RPCResponse, wsResponsesChannelCapacity),
+	}
+	wsClient.QuitService = *NewQuitService(log, "WSClient", wsClient)
+	return wsClient
+}
+
+func (wsc *WSClient) OnStart() error {
+	wsc.QuitService.OnStart()
+	err := wsc.dial()
+	if err != nil {
+		return err
+	}
+	go wsc.receiveEventsRoutine()
+	return nil
+}
+
+func (wsc *WSClient) dial() error {
+	// Dial
+	dialer := websocket.DefaultDialer
+	rHeader := http.Header{}
+	con, _, err := dialer.Dial(wsc.Address, rHeader)
+	if err != nil {
+		return err
+	}
+	wsc.Conn = con
+	return nil
+}
+
+func (wsc *WSClient) OnStop() {
+	wsc.QuitService.OnStop()
+}
+
+func (wsc *WSClient) receiveEventsRoutine() {
+	for {
+		_, data, err := wsc.ReadMessage()
+		if err != nil {
+			log.Info(Fmt("WSClient failed to read message: %v", err))
+			wsc.Stop()
+			break
+		} else {
+			var response rpctypes.RPCResponse
+			if err := json.Unmarshal(data, &response); err != nil {
```

Comment:
Because the result's data is an interface{}, and wire doesn't know how to unmarshal an (unregistered) interface{} type.

We can either register all the possible event Data types under a new interface type, or
we can make a special case for interface{} types and just return the underlying JSON object, in this case the map[string]interface{} thing.

Or, we can hack it with JSON as we did here.  I'm not sure what's best.